tensorflow/compiler/xla/service/gpu/horizontal_fusion.cc

```diff
@@ -0,0 +1,447 @@
+      VLOG(2) << "Reject unsupported fusion insn " << insn->ToString();
+      continue;
+    } else if (!is_consumer_the_only_user(*insn)) {
+      VLOG(2) << "Reject maybe illegal insn " << insn->ToString();
+      continue;
+    } else if (!is_profitable(*insn)) {
+      VLOG(2) << "Reject may-not-be profitable fusion insn "
+              << insn->ToString();
+      continue;
+    } else if (!all_row_major(*insn)) {
+      VLOG(2) << "Reject non-row-major fusion insn " << insn->ToString();
+      continue;
+    } else {
+      fusion_insns_.push_back(insn);
+    }
+  }
+
+  // Sort according to the number of outputs and instruciton counts.
+  std::sort(fusion_insns_.begin(), fusion_insns_.end(),
+            [&](const HloInstruction* a, const HloInstruction* b) {
+              if (GetOutputSizeOfFusion(*a) == GetOutputSizeOfFusion(*b)) {
+                return a->fused_instruction_count() <
+                       b->fused_instruction_count();
+              }
+              return GetOutputSizeOfFusion(*a) < GetOutputSizeOfFusion(*b);
+            });
+}
+
+// Get a next span of fusion instructions to be fused.
+absl::Span<HloInstruction*>
+HorizontalFusionImpl::FusionCandidates::GetNextSpanOfFusions() {
+  if (pos_ >= fusion_insns_.size()) {
+    return absl::Span<HloInstruction*>();
+  }
+
+  // Fusing too many computations at a time may not be easily profitable and
+  // may increase compile time due to large kernels. Set a limit to it.
+  constexpr int64 kMaxFusionBatchSize = 32;
+  // CUDA has a parameter size limit of ~4k bytes.
+  constexpr int64 kMaxCudaParamSize = 4000;
+  size_t accum_io_size = 0;
+  auto reach_max_fusion_batch_size = [&](size_t left, size_t right) -> bool {
+    if (right - left >= kMaxFusionBatchSize) {
+      return true;
+    }
+
+    accum_io_size += fusion_insns_.at(right)->fused_parameters().size() +
+                     GetOutputSizeOfFusion(*fusion_insns_.at(right));
+
+    if (accum_io_size * 8 >= kMaxCudaParamSize) {
+      return true;
+    }
+
+    return false;
+  };
+
+  size_t left = pos_;
+  size_t right = pos_ + 1;
+  size_t first_output_size = GetOutputSizeOfFusion(*fusion_insns_[left]);
+  for (; right < fusion_insns_.size(); ++right) {
+    if (first_output_size != GetOutputSizeOfFusion(*fusion_insns_[right])) {
+      // Cannot fuse computations who have different numbers of outputs.
+      break;
+    } else if (fusion_insns_[left]->fused_instruction_count() !=
+               fusion_insns_[right]->fused_instruction_count()) {
+      // Do not fuse computations of different instruction counts as it may
+      // introduce control divergence.
+      break;
+    } else if (reach_max_fusion_batch_size(left, right)) {
+      // Hit max fusion batch size.
+      break;
+    }
+  }
+
+  pos_ = right;
+  return absl::MakeSpan(fusion_insns_).subspan(left, right - left);
+}
+
+Status HorizontalFusionImpl::CreateFusedComputation(
+    absl::Span<HloInstruction*> fused_fusion_insns,
```

Comment:
Did you try MergeFusionInstructionIntoMultiOutput and friends? I know you need additional Reshapes and Slices, but maybe it's possible to add them after the merge.

Reply:
One difficulty could be that we need a mechanism to record the outputs of the fused computations (for later insertion of concat, etc). Using `MergeFusionInstructionIntoMultiOutput` may not be cleaner. Any thoughts on that?